src/java/main/org/apache/zookeeper/server/quorum/Leader.java

```diff
@@ -216,18 +218,35 @@ public boolean isQuorumSynced(QuorumVerifier qv) {
     
     private final ServerSocket ss;
 
-    Leader(QuorumPeer self,LeaderZooKeeperServer zk) throws IOException {
+    Leader(QuorumPeer self,LeaderZooKeeperServer zk) throws IOException, X509Exception {
         this.self = self;
         try {
-            if (self.getQuorumListenOnAllIPs()) {
-                ss = new ServerSocket(self.getQuorumAddress().getPort());
+            if (self.shouldUsePortUnification()) {
```

Comment:
This is a tough call IMO. I think you are correct that moving this out of the Leader would make the leader code simpler but on the other hand this is logic that I think "belongs" to the leader and is not really relevant anywhere else.

What do you think about the compromise of moving this to a static method in Leader where it can be tested and prevents cluttering up the constructor logic?